luarules/gadgets/unit_builder_priority.lua

```diff
@@ -15,49 +15,143 @@
+local lowPrioBuildPowerUsedRule = "lowPrioBuildPowerUsed" -- The total buildpower of constructors set to low priority that is actually spent on building stuff
+
+
+
 if not gadgetHandler:IsSyncedCode() then
-    return
-end
+	if DEBUG then 
+		
+		local canPassive = {} -- canPassive[unitDefID] = nil / true
+		for unitDefID, unitDef in pairs(UnitDefs) do
+			canPassive[unitDefID] = ((unitDef.canAssist and unitDef.buildSpeed > 0) or #unitDef.buildOptions > 0)
+		end
+		
+		function gadget:DrawWorld()
+			for i, unitID in pairs(Spring.GetAllUnits()) do 
+				if Spring.IsUnitInView(unitID) and canPassive[Spring.GetUnitDefID(unitID)] then  
+					local lowPriority = Spring.GetUnitRulesParam(unitID, ruleName)
+				
+					local x,y,z = Spring.GetUnitPosition(unitID)
+					gl.PushMatrix()
+					gl.Translate(x,y+64, z)
+					
+					gl.Text(((lowPriority and "L:"..tostring(lowPriority)) or "High"),0,0,16,'n')
+					gl.PopMatrix()
+
+				end
+			end
+			
+		end
+	end
+	if FORWARDUNIFORMS then 
+		local uniformCache = {0}
+		local function BuildSpeedsChanged(cmdname, countx2, ...)
+			local vararg = {...}
+			for i= 1, countx2, 2 do 
+				uniformCache[1] = vararg[i+1]
+				gl.SetUnitBufferUniforms(vararg[i], uniformCache, 1)
+				--Spring.Echo(vararg[i], vararg[i+1])
+			end
+			
+		end
+		
+		function gadget:Initialize()
+			gadgetHandler:AddSyncAction("BuildSpeedsChanged", BuildSpeedsChanged)
+		end
+
+		function gadget:Shutdown()
+			gadgetHandler:RemoveSyncAction("BuildSpeedsChanged")
+		end
+	end
+
+else
 
 local CMD_PRIORITY = 34571
 
-local stallMarginInc = 0.2
+local stallMarginInc = 0.2 -- 
 local stallMarginSto = 0.01
 
-local passiveCons = {} -- passiveCons[teamID][builderID]
-
-local buildTargets = {} --the unitIDs of build targets of passive builders
-local buildTargetOwners = {} --each build target has one passive builder that doesn't turn fully off, to stop the building decaying
+local buildPowerMinimum = 0.001 -- A small amount of buildpower we allocate to each build target owner to ensure that nanoframes dont vanish 
 
+local passiveCons = {} -- passiveCons[teamID][builderID] = true for passive cons
+local roundRobinIndexTeam = {} -- {teamID = roundRobinIndex}
+local roundRobinLastRoundTeamBuilders = {} -- {teamID = {builderID = true}} -- this is for taking away the resources of previous RR recipients
 local canBuild = {} --builders[teamID][builderID], contains all builders
-local realBuildSpeed = {} --build speed of builderID, as in UnitDefs (contains all builders)
-local currentBuildSpeed = {} --build speed of builderID for current interval, not accounting for buildOwners special speed (contains only passive builders)
 
-local costID = {} -- costID[unitID] (contains all units)
+local buildTargets = {} --{builtUnitID = builderUnitID} the unitIDs of build targets of passive builders, a -1 here indicates that this build target has Active builders too.
```

Comment:
This isn't actually being set to -1 for high-prio builders, so in some cases low-prio builders are being assigned 0.0001 BP (to keep nanoframes alive) despite a high-prio builder working on the same build. Fixed here: https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2600/commits/fc016e18a64174798598544c2a38bcaf0a044247

Reply:
Indeed, thats a leftover incorrectness on my part. Thank for catching it.

Reply:
I found that the added complexity complexity of clearing a buildtarget consistently if it had active builders was excessive compared to the buildpower allocation incorrectness cost.